### PR TITLE
sse implemented for service status and nat events

### LINKED ISSUE #735: SSE service for pushing updates to node consumer apps
We need a way to receive status changes, statistics, other data without polling via TequilApi since it's ineffective and requires a lot of boilerplate code.

AC:
  * [ ] Start WS server in node
  * [ ] Send service status
  * [ ] Send NAT status

### PR DESCRIPTION
**Pull request by vkuznecovas**
Closes #735 

This is ready to be merged, will improve testing in a separate PR since this is most likely blocking UI development.

### CODE REVIEW

File: core/service/pool.go [hunk 1]
```diff
@@ -173,5 +178,25 @@ func (i *Instance) Proposal() market.ServiceProposal {
 
 // State returns the service instance state.
 func (i *Instance) State() State {
+	i.stateLock.RLock()
+	defer i.stateLock.RUnlock()
 	return i.state
 }
+
+func (i *Instance) setState(newState State) {
```
Comment (tadovas) [inline core/service/pool.go]: It's a bit weird that state switch is not related to event publishing. Because this is exact place where events happen.  I would expect some kind of SetState or SwitchState which will publish even immediately after state switch.

File: tequilapi/endpoints/sse.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+// SSEHandler represents the sse handler
+type SSEHandler interface {
+	Sub(resp http.ResponseWriter, req *http.Request, params httprouter.Params)
+}
+
+// AddRoutesForSSE adds route for sse
+func AddRoutesForSSE(router *httprouter.Router, handler SSEHandler) {
+	router.GET("/sse", handler.Sub)
```
Comment (tadovas) [inline tequilapi/endpoints/sse.go]: SSE is content type, I would prefer more descriptive url like "/events", "/service-states" or something similar

File: core/service/manager.go [hunk 3]
```diff
@@ -145,14 +144,15 @@ func (manager *Manager) Start(providerID identity.Identity, serviceType string,
 	manager.servicePool.Add(&instance)
 
 	go func() {
-		instance.state = Running
+		instance.setState(Running)
+
+		manager.eventPublisher.Publish(StartTopic, instance.ToEvent())
```
Comment (Waldz) [inline core/service/manager.go]: Shouldn't event publish happen inside state change. It's also with lock

File: core/service/pool.go [hunk 4]
```diff
@@ -173,5 +178,25 @@ func (i *Instance) Proposal() market.ServiceProposal {
...
+	i.stateLock.Lock()
+	defer i.stateLock.Unlock()
+	i.state = newState
+}
+
+// ToEvent returns an event representation of the instance
+func (i *Instance) ToEvent() EventPayload {
+	i.stateLock.RLock()
+	defer i.stateLock.RUnlock()
+	return EventPayload{
```
Comment (Waldz) [inline core/service/pool.go]: I think we can send session instance, anyways it's for internal usage

File: core/service/pool.go (hunk 1 above)
Comment (vkuznecovas) [reply core/service/pool.go]: moved publish to setState

File: tequilapi/endpoints/sse.go (hunk 2 above)
Comment (vkuznecovas) [reply tequilapi/endpoints/sse.go]: renamed to /events

File: core/service/cleaner.go [hunk 5]
```diff
@@ -24,10 +24,12 @@ type Cleaner struct {
...
 // SessionStorage keeps sessions and allows removing them by proposal id
 type SessionStorage interface {
-	RemoveForService(serviceId string)
+	RemoveForService(serviceID string)
 }
 
 // Cleanup removes sessions of stopped service
-func (cleaner *Cleaner) Cleanup(instance *Instance) {
-	cleaner.SessionStorage.RemoveForService(string(instance.id))
+func (cleaner *Cleaner) Cleanup(event EventPayload) {
```
Comment (zolia) [inline core/service/cleaner.go]: Its probably not a `Cleanup` then, since imho its should be called only once and we don't care about event's state.
